### DISCUSSION TITLE
[Feature]:  prevent clipPath from clipping setBackgroundImage

### DISCUSSION THREAD

**Discussion by mezzouhri** [fabricjs/fabric.js#8895] [Q&A] [answered]
### CheckList

- [X] I agree to follow this project's Code of Conduct
- [X] I have read and followed the [Contributing Guide](https://github.com/fabricjs/fabric.js/blob/master/CONTRIBUTING.md)
- [X] I have searched and referenced existing issues, feature requests and discussions
- [X] I am filing a **FEATURE** request.

### Description

Hi,
Could you please tell me how to prevent canvas.clipPath from clipping canvas.setBackgroundImage,
because when I apply canvas.clipPath, the clipPath clip also the backgroundImage : 

here what happens : 
<img width="1680" alt="Screenshot 2023-04-11 at 15 49 40" src="https://user-images.githubusercontent.com/24857043/231184521-57891534-8156-437e-b660-3ec4c26f2eec.png">
my code : 

    const points = [
      { x: 0, y: -302.36 },
      { x: 302.36, y: -151.18 },
      { x: 302.36, y: 151.18 },
      { x: 0, y: 302.36 },
      { x: -302.36, y: 151.18 },
      { x: -302.36, y: -151.18 },
    ];
    canvas.clipPath = new fabric.Polygon(points, {
      originX: "center",
      originY: "center",
    });
    newCanvas.setBackgroundImage(
      "images/overlayImage.svg",
      newCanvas.renderAll.bind(newCanvas),
      {
        originX: "center",
        originY: "center",
        fill: "transparent",
        selectable: false,
        evented: false,
        hasBorders: false,
        hasControls: false,
        hasRotatingPoint: false,
        excludeFromExport: true,
      }
    );
    

What I want : 
<img width="1654" alt="image" src="https://user-images.githubusercontent.com/24857043/231185266-70fc6015-b6d2-4e8d-be1f-be54d6dee044.png">

I can achieve this behavior by using `setOverlayImage` instead of `setBackgroundImage` (I want to use setBackgroundImage)

    const points = [
      { x: 0, y: -302.36 },
      { x: 302.36, y: -151.18 },
      { x: 302.36, y: 151.18 },
      { x: 0, y: 302.36 },
      { x: -302.36, y: 151.18 },
      { x: -302.36, y: -151.18 },
    ];
    canvas.clipPath = new fabric.Polygon(points, {
      originX: "center",
      originY: "center",
    });
    newCanvas.setOverlayImage(
      "images/overlayImage.svg",
      newCanvas.renderAll.bind(newCanvas),
      {
        originX: "center",
        originY: "center",
        fill: "transparent",
        selectable: false,
        evented: false,
        hasBorders: false,
        hasControls: false,
        hasRotatingPoint: false,
        excludeFromExport: true,
      }
    );

    
Thank you 

### Current State

current State : the clipPath clip background image

**Comment by mezzouhri** [comment, marked as answer]
Hi, I have found a solution for this problem, I don't know if it's optimal or not, but it works fine : 
I share it with you in case if someone need the same behavior : 
```
//Add polygon as clipPath
    const points = [
      { x: 0, y: -302.36 },
      { x: 302.36, y: -151.18 },
      { x: 302.36, y: 151.18 },
      { x: 0, y: 302.36 },
      { x: -302.36, y: 151.18 },
      { x: -302.36, y: -151.18 },
    ];

    //canvas (Background Image) clipPath
    const clipPath1 = new fabric.Polygon(points, {
      originX: "center",
      originY: "center",
    });

    //Overlay Image clipPath
    const clipPath2 = new fabric.Polygon(points, {
      originX: "center",
      originY: "center",
      inverted: true,
    });

    canvas.clipPath = clipPath1;

    canvas.setBackgroundImage(
      "images/my-image.svg", // path to image
      canvas.renderAll.bind(canvas),
      {
        originX: "center",
        originY: "center",
        fill: "transparent",
        selectable: false,
        evented: false,
        hasBorders: false,
        hasControls: false,
        hasRotatingPoint: false,
        excludeFromExport: true,
      }
    );
    canvas.setOverlayImage(
      "images/my-image.svg",
      canvas.renderAll.bind(canvas),
      {
        originX: "center",
        originY: "center",
        fill: "transparent",
        selectable: false,
        evented: false,
        hasBorders: false,
        hasControls: false,
        hasRotatingPoint: false,
        excludeFromExport: true,
        clipPath: clipPath2,
      }
    );
```

**Reply by pawan-everbee** [reply]
hey @mezzouhri can you please explain what is happening here. I have the similar problem right now, i have a canvas of 450 by 450 px and i am adding a rectangle as clippath now i want to add a background image to the whole canvas (450*450) but the image is getting clipped inside the clippath

**Reply by mezzouhri** [reply]
Hi @pawan-everbee, when you add a clipPath to canvas, this one clip also the background image.
if you want add a background image to the whole canvas, you could use my solution proposed above.

I've created 2 clipPaths : normal and inverted clipPath, you can refer to this link to understand how inverted clipPath works  [http://fabricjs.com/clippath-part4](http://fabricjs.com/clippath-part4)

then, I've used 2 images :
 - first image as a background image only inside the rectangle using the normal clipPath 
 - second image as a overlay image outside the rectangle using the inverted clipPath 
 
 please tell me if it's ok for you or not.

**Reply by pawan-everbee** [reply]
@mezzouhri thanks for replying but my situation is bit different i think as i added the clippath rectangle in the center of the canvas and it's transparent, now i want to add the backround image to the whole canvas so it's visible behind the clippath as well inside the whole canvas

**Reply by mezzouhri** [reply]
@pawan-everbee , no, I think is exactly the same situation, could you create a sandcodebox or any online example so I can help you ?
with the same image you should use :
 - `canvas.setBackgroundImage` with normal clipPath1
  - `canvas.setOverlayImage` with normal clipPath2 = inverted clipPath1,

**Reply by pawan-everbee** [reply]
@mezzouhri here is a working jsfiddle for my problem, i have added a clippath in the center and then added one background image and still being restricted inside the clippath itself
http://jsfiddle.net/fjyLh9v0/

**Reply by pawan-everbee** [reply]
@mezzouhri i thank you for taking your time to reply, i have attached the updated fiddle can you please let me know if there is anything which needs to be changed, any help is appreciated.

**Reply by ahmedrizwan239** [reply]
Hi @mezzouhri , i think the fiddle you attached doesn't work as expected, it creates issues in selecting shapes and releasing the controls. Can you please check the sandbox URL I shared? I've created a box and i want to clip everything outside that except for the background image. 

Sandbox URL:
https://stackblitz.com/edit/sb1-o7rrpp8b?file=index.html

Discussion thread:
https://github.com/fabricjs/fabric.js/discussions/10579

**Comment by mezzouhri** [comment]
Hi @pawan-everbee , here is [a quick example](https://codesandbox.io/s/bold-glitter-7kwqlj?file=/src/components/Canvas/index.jsx) I created on codesandbox hundle your problem
